src/Processors/Transforms/PartialSortingTransform.cpp

```diff
@@ -55,33 +81,39 @@ void PartialSortingTransform::transform(Chunk & chunk)
     ColumnRawPtrs block_columns;
     UInt64 rows_num = block.rows();
 
+    /** If we've saved columns from previously blocks we could filter all rows from current block
+      * which are unnecessary for sortBlock(...) because they obviously won't be in the top LIMIT rows.
+      */
     if (!threshold_block_columns.empty())
     {
-        IColumn::Filter filter(rows_num, 0);
         block_columns = extractColumns(block, description);
-        size_t filtered_count = 0;
-
-        for (UInt64 i = 0; i < rows_num; ++i)
-        {
-            if (less(threshold_block_columns, limit - 1, block_columns, i, description))
-            {
-                ++filtered_count;
-                filter[i] = 1;
-            }
-        }
+        IColumn::Filter filter;
+        size_t filtered_count = getFilterMask(block_columns, threshold_block_columns, limit - 1, description, rows_num, filter);
 
         if (filtered_count)
         {
+            auto expected_size = rows_num - filtered_count;
+            size_t i = 0;
+            Columns new_columns(block.columns());
             for (auto & column : block.getColumns())
             {
-                column = column->filter(filter, filtered_count);
+                new_columns[i++] = column->filter(filter, expected_size);
             }
+            block.setColumns(new_columns);
         }
     }
 
     sortBlock(block, description, limit);
 
-    if (limit && limit < block.rows() &&
+    if (!threshold_block_columns.empty())
+    {
+        block_columns = extractColumns(block, description);
+    }
+
+    /** If this is the first processed block or (limit - 1)'th row of the current block
+      * is less than current threshold row then we could update threshold.
+      */
+    if (1500 <= limit && limit <= block.rows() &&
```

Comment:
If we use `1500` as a constant, let's give it a name and add a comment why do we chose this value.